File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/VoteData/LocalVoteDataSource.java
package com.heaton.funnyvote.data.VoteData;

import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.heaton.funnyvote.data.local.dao.OptionDao;
import com.heaton.funnyvote.data.local.dao.VoteDataDao;
import com.heaton.funnyvote.database.Option;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;
import com.heaton.funnyvote.utils.AppExecutors;

import java.io.File;
import java.util.List;

public class LocalVoteDataSource implements VoteDataSource {
    private static final String TAG = LocalVoteDataSource.class.getSimpleName();
    private VoteDataDao voteDataDao;
    private OptionDao optionDao;
    private static volatile LocalVoteDataSource INSTANCE;
    private AppExecutors mAppExecutors;

    public LocalVoteDataSource(@NonNull VoteDataDao voteDataDao, OptionDao optionDao, AppExecutors appExecutors) {
        this.voteDataDao = voteDataDao;
        this.optionDao = optionDao;
        this.mAppExecutors = appExecutors;
    }

    public static LocalVoteDataSource getInstance(@NonNull VoteDataDao voteDataDao,
                                                  @NonNull OptionDao optionDao,
                                                  AppExecutors appExecutors) {
        if (INSTANCE == null) {
            synchronized (LocalVoteDataSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new LocalVoteDataSource(voteDataDao, optionDao, appExecutors);
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public void getVoteData(final String voteCode, User user, @Nullable final GetVoteDataCallback callback) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                if (TextUtils.isEmpty(voteCode)) {
                    if (callback != null) callback.onVoteDataNotAvailable();
                    return;
                }
                final VoteData voteData = voteDataDao.getVoteByCode(voteCode);
                if (voteData != null) {
                    List<Option> options = optionDao.getOptionsByVoteCode(voteCode);
                    voteData.setOptions(options);
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            if (callback != null) callback.onVoteDataLoaded(voteData);
                        }
                    });
                } else {
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            if (callback != null) callback.onVoteDataNotAvailable();
                        }
                    });
                }
            }
        });
    }

    @Override
    public void saveVoteData(final VoteData voteDataNetwork) {
        List<Option> optionList = voteDataNetwork.getNetOptions();
        if (optionList != null) {
            voteDataNetwork.setOptionCount(optionList.size());
            int maxOption = 0;
            for (int i = 0; i < optionList.size(); i++) {
                Option option = optionList.get(i);
                option.setVoteCode(voteDataNetwork.getVoteCode());
                if (option.getCount() == null) {
                    option.setCount(0);
                }
                if (i == 0) {
                    voteDataNetwork.setOption1Title(option.getTitle());
                    voteDataNetwork.setOption1Code(option.getCode());
                    voteDataNetwork.setOption1Count(option.getCount());
                    voteDataNetwork.setOption1Polled(option.getIsUserChoiced());
                } else if (i == 1) {
                    voteDataNetwork.setOption2Title(option.getTitle());
                    voteDataNetwork.setOption2Code(option.getCode());
                    voteDataNetwork.setOption2Count(option.getCount());
                    voteDataNetwork.setOption2Polled(option.getIsUserChoiced());
                }
                if (option.getCount() > maxOption && option.getCount() >= 1) {
                    maxOption = option.getCount();
                    voteDataNetwork.setOptionTopCount(option.getCount());
                    voteDataNetwork.setOptionTopCode(option.getCode());
                    voteDataNetwork.setOptionTopTitle(option.getTitle());
                    voteDataNetwork.setOptionTopPolled(option.getIsUserChoiced());
                }
                if (option.getIsUserChoiced()) {
                    voteDataNetwork.setOptionUserChoiceCode(option.getCode());
                    voteDataNetwork.setOptionUserChoiceTitle(option.getTitle());
                    voteDataNetwork.setOptionUserChoiceCount(option.getCount());
                }
            }
        }
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                voteDataDao.deleteByCode(voteDataNetwork.getVoteCode());
                optionDao.deleteByVoteCode(voteDataNetwork.getVoteCode());
                voteDataDao.insert(voteDataNetwork);
                if (voteDataNetwork.getNetOptions() != null) {
                    for (Option opt : voteDataNetwork.getNetOptions()) {
                        opt.setVoteCode(voteDataNetwork.getVoteCode());
                    }
                    optionDao.insertAll(voteDataNetwork.getNetOptions());
                }
            }
        });
    }

    @Override
    public void getOptions(final VoteData voteData, final GetVoteOptionsCallback callback) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                final List<Option> optionList = optionDao.getOptionsByVoteCode(voteData.getVoteCode());
                mAppExecutors.mainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        if (optionList != null && optionList.size() >= 2) {
                            callback.onVoteOptionsLoaded(optionList);
                        } else {
                            callback.onVoteOptionsNotAvailable();
                        }
                    }
                });
            }
        });
    }

    @Override
    public void saveOptions(final List<Option> optionList) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                optionDao.insertAll(optionList);
            }
        });
    }

    @Override
    public void saveVoteDataList(final List<VoteData> voteDataList, final int offset, final String tab) {
        for (int i = 0; i < voteDataList.size(); i++) {
            VoteData voteData = voteDataList.get(i);
            if (voteData.getFirstOption() != null) {
                voteData.setOption1Code(voteData.getFirstOption().getCode());
                voteData.setOption1Title(voteData.getFirstOption().getTitle());
                voteData.setOption1Count(voteData.getFirstOption().getCount());
                voteData.setOption1Polled(voteData.getFirstOption().getIsUserChoiced());
            }
            if (voteData.getSecondOption() != null) {
                voteData.setOption2Code(voteData.getSecondOption().getCode());
                voteData.setOption2Title(voteData.getSecondOption().getTitle());
                voteData.setOption2Count(voteData.getSecondOption().getCount());
                voteData.setOption2Polled(voteData.getSecondOption().getIsUserChoiced());
            }
            if (voteData.getTopOption() != null) {
                voteData.setOptionTopCode(voteData.getTopOption().getCode());
                voteData.setOptionTopTitle(voteData.getTopOption().getTitle());
                voteData.setOptionTopCount(voteData.getTopOption().getCount());
                voteData.setOptionTopPolled(voteData.getTopOption().getIsUserChoiced());
            }
            if (voteData.getUserOption() != null) {
                voteData.setOptionUserChoiceCode(voteData.getUserOption().getCode());
                voteData.setOptionUserChoiceTitle(voteData.getUserOption().getTitle());
                voteData.setOptionUserChoiceCount(voteData.getUserOption().getCount());
            }
            if (tab != null && tab.equals(VoteDataRepository.TAB_HOT)) {
                voteData.setDisplayOrder((offset) * VoteDataRepository.PAGE_COUNT + i);
                voteData.setCategory("hot");
            } else {
                voteData.setCategory(null);
            }
        }
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                for (VoteData data : voteDataList) {
                    voteDataDao.deleteByCode(data.getVoteCode());
                }
                voteDataDao.insertOrReplaceInTx(voteDataList);
            }
        });
    }

    @Override
    public void addNewOption(String voteCode, String password, List<String> newOptions, User user, AddNewOptionCallback callback) {
    }

    @Override
    public void pollVote(@NonNull String voteCode, String password, @NonNull List<String> pollOptions, @NonNull User user, @Nullable PollVoteCallback callback) {
    }

    @Override
    public void favoriteVote(final String voteCode, final boolean isFavorite, User user, final FavoriteVoteCallback callback) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                VoteData voteData = voteDataDao.getVoteByCode(voteCode);
                if (voteData != null) {
                    voteData.setIsFavorite(isFavorite);
                    voteDataDao.update(voteData);
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            if (callback != null) callback.onSuccess(isFavorite);
                        }
                    });
                }
            }
        });
    }

    @Override
    public void createVote(@NonNull VoteData voteSetting, @NonNull List<String> options, File image, GetVoteDataCallback callback) {
    }

    @Override
    public void getHotVoteList(final int offset, User user, final GetVoteListCallback callback) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                final List<VoteData> list = voteDataDao.getHotVotes("hot", VoteDataRepository.PAGE_COUNT, offset);
                mAppExecutors.mainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        callback.onVoteListLoaded(list);
                    }
                });
            }
        });
    }

    @Override
    public void getCreateVoteList(final int offset, final User user, User targetUser, final GetVoteListCallback callback) {
        if (targetUser == null && user != null) {
            mAppExecutors.diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    final List<VoteData> list = voteDataDao.getCreateVotes(user.getUserCode(), VoteDataRepository.PAGE_COUNT, offset);
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            callback.onVoteListLoaded(list);
                        }
                    });
                }
            });
        } else {
            callback.onVoteListNotAvailable();
        }
    }

    @Override
    public void getParticipateVoteList(final int offset, final User user, User targetUser, final GetVoteListCallback callback) {
        if (targetUser == null) {
            mAppExecutors.diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    final List<VoteData> list = voteDataDao.getParticipateVotes(VoteDataRepository.PAGE_COUNT, offset);
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            callback.onVoteListLoaded(list);
                        }
                    });
                }
            });
        } else {
            callback.onVoteListNotAvailable();
        }
    }

    @Override
    public void getFavoriteVoteList(final int offset, User user, User targetUser, final GetVoteListCallback callback) {
        if (targetUser == null) {
            mAppExecutors.diskIO().execute(new Runnable() {
                @Override
                public void run() {
                    final List<VoteData> list = voteDataDao.getFavoriteVotes(VoteDataRepository.PAGE_COUNT, offset);
                    mAppExecutors.mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            callback.onVoteListLoaded(list);
                        }
                    });
                }
            });
        } else {
            callback.onVoteListNotAvailable();
        }
    }

    @Override
    public void getSearchVoteList(final String keyword, final int offset, @NonNull User user, final GetVoteListCallback callback) {
        if (TextUtils.isEmpty(keyword)) {
            callback.onVoteListNotAvailable();
            return;
        }
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                final List<VoteData> list = voteDataDao.searchVotes(keyword, VoteDataRepository.PAGE_COUNT, offset);
                mAppExecutors.mainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        callback.onVoteListLoaded(list);
                    }
                });
            }
        });
    }

    @Override
    public void getNewVoteList(final int offset, User user, final GetVoteListCallback callback) {
        mAppExecutors.diskIO().execute(new Runnable() {
            @Override
            public void run() {
                final List<VoteData> list = voteDataDao.getNewVotes(VoteDataRepository.PAGE_COUNT, offset);
                mAppExecutors.mainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        callback.onVoteListLoaded(list);
                    }
                });
            }
        });
    }

    @VisibleForTesting
    static void clearInstance() {
        INSTANCE = null;
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/VoteData/LocalVoteDataSource.kt
package com.heaton.funnyvote.data.VoteData

import android.text.TextUtils
import androidx.annotation.VisibleForTesting
import com.heaton.funnyvote.data.local.dao.OptionDao
import com.heaton.funnyvote.data.local.dao.VoteDataDao
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData
import com.heaton.funnyvote.ui.main.MainPageTabFragment
import com.heaton.funnyvote.utils.AppExecutors
import java.io.File

class LocalVoteDataSource private constructor(
    private val voteDataDao: VoteDataDao,
    private val optionDao: OptionDao,
    private val mAppExecutors: AppExecutors
) : VoteDataSource {

    override fun getVoteData(voteCode: String, user: User, callback: VoteDataSource.GetVoteDataCallback?) {
        mAppExecutors.diskIO.execute {
            if (TextUtils.isEmpty(voteCode)) {
                mAppExecutors.mainThread.execute { callback?.onVoteDataNotAvailable() }
                return@execute
            }
            val voteData = voteDataDao.getVoteByCode(voteCode)
            if (voteData != null) {
                val options = optionDao.getOptionsByVoteCode(voteCode)
                voteData.options = options
                mAppExecutors.mainThread.execute {
                    callback?.onVoteDataLoaded(voteData)
                }
            } else {
                mAppExecutors.mainThread.execute {
                    callback?.onVoteDataNotAvailable()
                }
            }
        }
    }

    override fun saveVoteData(voteData: VoteData) {
        val optionList = voteData.netOptions ?: emptyList()
        voteData.optionCount = optionList.size
        var maxOption = 0
        for (i in optionList.indices) {
            val option = optionList[i]
            option.voteCode = voteData.voteCode
            if (option.count == null) {
                option.count = 0
            }
            option.id = null
            if (i == 0) {
                voteData.option1Title = option.title
                voteData.option1Code = option.code
                voteData.option1Count = option.count
                voteData.option1Polled = option.isUserChoiced
            } else if (i == 1) {
                voteData.option2Title = option.title
                voteData.option2Code = option.code
                voteData.option2Count = option.count
                voteData.option2Polled = option.isUserChoiced
            }
            if (option.count > maxOption && option.count >= 1) {
                maxOption = option.count
                voteData.optionTopCount = option.count
                voteData.optionTopCode = option.code
                voteData.optionTopTitle = option.title
                voteData.optionTopPolled = option.isUserChoiced
            }
            if (option.isUserChoiced) {
                voteData.optionUserChoiceCode = option.code
                voteData.optionUserChoiceTitle = option.title
                voteData.optionUserChoiceCount = option.count
            }
        }
        mAppExecutors.diskIO.execute {
            voteDataDao.deleteByCode(voteData.voteCode)
            voteDataDao.insertOrReplace(voteData)
            optionDao.deleteByVoteCode(voteData.voteCode)
            optionDao.insertOrReplaceInTx(optionList)
        }
    }

    override fun getOptions(voteData: VoteData, callback: VoteDataSource.GetVoteOptionsCallback) {
        mAppExecutors.diskIO.execute {
            val optionList = optionDao.getOptionsByVoteCode(voteData.voteCode)
            mAppExecutors.mainThread.execute {
                if (optionList.size >= 2) {
                    callback.onVoteOptionsLoaded(optionList)
                } else {
                    callback.onVoteOptionsNotAvailable()
                }
            }
        }
    }

    override fun saveOptions(optionList: List<Option>) {
        mAppExecutors.diskIO.execute {
            optionDao.insertOrReplaceInTx(optionList)
        }
    }

    override fun saveVoteDataList(voteDataList: List<VoteData>, offset: Int, tab: String) {
        for (i in voteDataList.indices) {
            val voteData = voteDataList[i]
            if (voteData.firstOption != null) {
                voteData.option1Code = voteData.firstOption.code
                voteData.option1Title = voteData.firstOption.title
                voteData.option1Count = voteData.firstOption.count
                voteData.option1Polled = voteData.firstOption.isUserChoiced
            }
            if (voteData.secondOption != null) {
                voteData.option2Code = voteData.secondOption.code
                voteData.option2Title = voteData.secondOption.title
                voteData.option2Count = voteData.secondOption.count
                voteData.option2Polled = voteData.secondOption.isUserChoiced
            }
            if (voteData.topOption != null) {
                voteData.optionTopCode = voteData.topOption.code
                voteData.optionTopTitle = voteData.topOption.title
                voteData.optionTopCount = voteData.topOption.count
                voteData.optionTopPolled = voteData.topOption.isUserChoiced
            }
            if (voteData.userOption != null) {
                voteData.optionUserChoiceCode = voteData.userOption.code
                voteData.optionUserChoiceTitle = voteData.userOption.title
                voteData.optionUserChoiceCount = voteData.userOption.count
            }
            if (tab == MainPageTabFragment.TAB_HOT) {
                voteData.displayOrder = offset * VoteDataRepository.PAGE_COUNT + i
                voteData.category = "hot"
            } else {
                voteData.category = null
            }
        }
        mAppExecutors.diskIO.execute {
            for (data in voteDataList) {
                voteDataDao.deleteByCode(data.voteCode)
            }
            voteDataDao.insertOrReplaceInTx(voteDataList)
        }
    }

    override fun addNewOption(voteCode: String, password: String, newOptions: List<String>, user: User, callback: VoteDataSource.AddNewOptionCallback) {}

    override fun pollVote(voteCode: String, password: String, pollOptions: List<String>, user: User, callback: VoteDataSource.PollVoteCallback?) {}

    override fun favoriteVote(voteCode: String, isFavorite: Boolean, user: User, callback: VoteDataSource.FavoriteVoteCallback) {
        mAppExecutors.diskIO.execute {
            val voteData = voteDataDao.getVoteByCode(voteCode)
            if (voteData != null) {
                voteData.isFavorite = isFavorite
                voteDataDao.update(voteData)
                mAppExecutors.mainThread.execute {
                    callback.onSuccess(isFavorite)
                }
            } else {
                mAppExecutors.mainThread.execute {
                    callback.onFailure()
                }
            }
        }
    }

    override fun createVote(voteSetting: VoteData, options: List<String>, image: File?, callback: VoteDataSource.GetVoteDataCallback) {}

    override fun getHotVoteList(offset: Int, user: User, callback: VoteDataSource.GetVoteListCallback) {
        mAppExecutors.diskIO.execute {
            val list = voteDataDao.getHotVotes("hot", VoteDataRepository.PAGE_COUNT, offset)
            mAppExecutors.mainThread.execute {
                callback.onVoteListLoaded(list)
            }
        }
    }

    override fun getCreateVoteList(offset: Int, loginUser: User, targetUser: User, callback: VoteDataSource.GetVoteListCallback) {
        if (targetUser.userCode.isNullOrEmpty() && !loginUser.userCode.isNullOrEmpty()) {
            mAppExecutors.diskIO.execute {
                val list = voteDataDao.getCreateVotes(loginUser.userCode, VoteDataRepository.PAGE_COUNT, offset)
                mAppExecutors.mainThread.execute {
                    callback.onVoteListLoaded(list)
                }
            }
        } else {
            callback.onVoteListNotAvailable()
        }
    }

    override fun getParticipateVoteList(offset: Int, loginUser: User, targetUser: User, callback: VoteDataSource.GetVoteListCallback) {
        if (targetUser.userCode.isNullOrEmpty()) {
            mAppExecutors.diskIO.execute {
                val list = voteDataDao.getParticipateVotes(VoteDataRepository.PAGE_COUNT, offset)
                mAppExecutors.mainThread.execute {
                    callback.onVoteListLoaded(list)
                }
            }
        } else {
            callback.onVoteListNotAvailable()
        }
    }

    override fun getFavoriteVoteList(offset: Int, loginUser: User, targetUser: User, callback: VoteDataSource.GetVoteListCallback) {
        if (targetUser.userCode.isNullOrEmpty()) {
            mAppExecutors.diskIO.execute {
                val list = voteDataDao.getFavoriteVotes(VoteDataRepository.PAGE_COUNT, offset)
                mAppExecutors.mainThread.execute {
                    callback.onVoteListLoaded(list)
                }
            }
        } else {
            callback.onVoteListNotAvailable()
        }
    }

    override fun getSearchVoteList(keyword: String, offset: Int, user: User, callback: VoteDataSource.GetVoteListCallback) {
        if (TextUtils.isEmpty(keyword)) {
            callback.onVoteListNotAvailable()
            return
        }
        mAppExecutors.diskIO.execute {
            val list = voteDataDao.searchVotes(keyword, VoteDataRepository.PAGE_COUNT, offset)
            mAppExecutors.mainThread.execute {
                callback.onVoteListLoaded(list)
            }
        }
    }

    override fun getNewVoteList(offset: Int, user: User, callback: VoteDataSource.GetVoteListCallback) {
        mAppExecutors.diskIO.execute {
            val list = voteDataDao.getNewVotes(VoteDataRepository.PAGE_COUNT, offset)
            mAppExecutors.mainThread.execute {
                callback.onVoteListLoaded(list)
            }
        }
    }

    companion object {
        private var INSTANCE: LocalVoteDataSource? = null

        @JvmStatic
        fun getInstance(
            voteDataDao: VoteDataDao,
            optionDao: OptionDao,
            appExecutors: AppExecutors
        ): LocalVoteDataSource {
            if (INSTANCE == null) {
                synchronized(LocalVoteDataSource::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = LocalVoteDataSource(voteDataDao, optionDao, appExecutors)
                    }
                }
            }
            return INSTANCE!!
        }

        @JvmStatic
        @VisibleForTesting
        fun clearInstance() {
            INSTANCE = null
        }
    }
}
